### ISSUE TITLE
Cache causes `dune build` to fail when `HEAD` doesn't exist

### ISSUE THREAD

**Issue by CraigFe** [ocaml/dune#4429]
The Dune cache uses `Vcs.commit_id` as an input to the Dune cache [here](https://github.com/ocaml/dune/blob/65172b313d6dd937fcec8ff5b59bac90cac94afd/src/dune_engine/build_system.ml#L436). Unfortunately, this commit ID might not always exist: when using Git, `HEAD` doesn't exist in commit-less repositories:

```bash
  $ git init

  Initialized empty Git repository in $TESTCASE_ROOT/.git/

  $ $ dune subst 2>&1 | sed 's/\/tmp\/.*.output/<tmp_file>/g'
           git (internal) (exit 128)
  /usr/bin/git describe --always --dirty > <tmp_file>
  fatal: bad revision 'HEAD'
           git (internal) (exit 128)
  /usr/bin/git rev-parse HEAD > <tmp_file>
  fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
  Use '--' to separate paths from revisions, like this:
  'git <command> [<revision>...] -- [<file>...]'
           git (internal) (exit 128)
  /usr/bin/git ls-tree -r --name-only HEAD > <tmp_file>
  fatal: Not a valid object name HEAD
```

As of today, this is also triggerable via `dune build` with the cache enabled:

```bash
  $ DUNE_CACHE=enabled dune build

           git (internal) (exit 128)
  /usr/bin/git rev-parse HEAD > /tmp/buildd528b2.dune/dune6c093a.output
  fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
  Use '--' to separate paths from revisions, like this:
  'git <command> [<revision>...] -- [<file>...]'
  [1]
```

As discussed in https://github.com/ocaml/dune/pull/4430, the latter case is going away for unrelated reasons.

Obviously only a tiny problem, but seemed worth a bug report.

## Reproduction

See https://github.com/ocaml/dune/pull/4430.

**Comment by emillon** [comment]
Duplicate of #3619

**Comment by emillon** [comment]
Thanks for the report! I found that before but it seemed innocuous - the fact that it can be triggered through the cache makes it a bit more important.

**Comment by CraigFe** [comment]
Ah, apologies for the duplicate; should have looked a bit more closely.

As @jeremiedimino and @snowleopard point out in the linked issue, the caching reproduction is likely to go away due to unrelated refactoring, so this isn't such a big problem. It is also possible to observe it via `dune-build-info`, however.

**Comment by rgrinberg** [comment]
This was fixed in #4441
